hazelcast-sql/src/test/java/com/hazelcast/jet/sql_nightly/SqlSTSInnerEquiJoinFaultToleranceStressTest.java

```diff
@@ -0,0 +1,287 @@
+    private SqlService sqlService;
+    private Thread kafkaFeedThread;
+    private String sourceTopicName;
+    private String fetchingQuery;
+    protected String sinkTopic;
+
+    protected int expectedEventsCount = EVENTS_TO_PROCESS;
+    protected int firstItemId = 1;
+    protected int lastItemId = EVENTS_TO_PROCESS;
+
+    @Parameter(value = 0)
+    public String processingGuarantee;
+
+    @Parameter(value = 1)
+    public boolean restartGraceful;
+
+    @Parameterized.Parameters(name = "processingGuarantee:{0}, restartGraceful:{1}")
+    public static Collection<Object[]> parameters() {
+        return asList(new Object[][]{
+                {AT_LEAST_ONCE, true},
+                {AT_LEAST_ONCE, false},
+                {EXACTLY_ONCE, true},
+                {EXACTLY_ONCE, false}
+        });
+    }
+
+    @BeforeClass
+    public static void beforeClass() throws IOException {
+        initialize(5, null);
+        kafkaTestSupport = KafkaTestSupport.create();
+        kafkaTestSupport.createKafkaCluster();
+    }
+
+    @AfterClass
+    public static void afterClass() {
+        kafkaTestSupport.shutdownKafkaCluster();
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        sqlService = instance().getSql();
+
+        // Kafka source definition
+        sourceTopicName = createRandomTopic();
+        sqlService.execute("CREATE MAPPING " + sourceTopicName + ' '
+                + "TYPE " + KafkaSqlConnector.TYPE_NAME + ' '
+                + "OPTIONS ( "
+                + '\'' + OPTION_KEY_FORMAT + "'='int'"
+                + ", '" + OPTION_VALUE_FORMAT + "'='varchar'"
+                + ", 'bootstrap.servers'='" + kafkaTestSupport.getBrokerConnectionString() + '\''
+                + ", 'auto.offset.reset'='earliest'"
+                + ")");
+
+        kafkaFeedThread = new Thread(() -> createTopicData(sqlService, sourceTopicName));
+        kafkaFeedThread.start();
+
+        // Kafka sink topic definition
+        sinkTopic = randomName();
+        sqlService.execute("CREATE MAPPING " + sinkTopic
+                + " TYPE " + KafkaSqlConnector.TYPE_NAME + ' '
+                + " OPTIONS ( "
+                + '\'' + OPTION_KEY_FORMAT + "'='int'"
+                + ", '" + OPTION_VALUE_FORMAT + "'='varchar'"
+                + ", 'bootstrap.servers'='" + kafkaTestSupport.getBrokerConnectionString() + '\''
+                + ", 'auto.offset.reset'='earliest'"
+                + ")");
+
+        // Left & right sides of the JOIN
+        sqlService.execute("CREATE VIEW s1 AS " +
+                "SELECT * FROM TABLE(IMPOSE_ORDER(TABLE " + sourceTopicName + " , DESCRIPTOR(__key), 10))");
+        sqlService.execute("CREATE VIEW s2 AS " +
+                "SELECT * FROM TABLE(IMPOSE_ORDER(TABLE " + sourceTopicName + " , DESCRIPTOR(__key), 5))");
+
+        fetchingQuery = setupFetchingQuery();
+
+        Thread.sleep(15_000L); // time to feed Kafka
```

Comment:
This is notoriously unreliable in busy test environment. If we're waiting for kafka production to finish, why do we start it in another thread? Do it in this thread and continue right after.